admin/server/github.go

```diff
@@ -88,53 +134,254 @@ func (s *Server) githubConnect(w http.ResponseWriter, r *http.Request, pathParam
+	http.Redirect(w, r, redirectURL.String(), http.StatusTemporaryRedirect)
 }
 
 // githubConnectCallback is called after a Github App authorization flow initiated by githubConnect has completed.
+// This call can originate from users who are not logged in in cases like admin user accepting installation request, removing existing installation etc.
 // It's implemented as a non-gRPC endpoint mounted directly on /github/connect/callback.
+// High level flow:
+// User installation is accompanied with user authorisation as well
+//   - verify the user is a collaborator else return unauthorised
+//   - verify the user installed the app on the right repo else navigate to retry
+//   - Save user's github username in the users table
+//
+// If user requests the app
+//   - navigate the user to requested page
 func (s *Server) githubConnectCallback(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
-	// TODO: Enable user authorization and verify user per https://roadie.io/blog/avoid-leaking-github-org-data/
+	ctx := r.Context()
 
 	// Extract info from query string
 	qry := r.URL.Query()
 	setupAction := qry.Get("setup_action")
-	if setupAction != "install" && setupAction != "update" { // TODO: Also handle "request"
+	if setupAction != "install" && setupAction != "update" && setupAction != "request" {
 		http.Error(w, fmt.Sprintf("unexpected setup_action=%q", setupAction), http.StatusBadRequest)
 		return
 	}
-	installationIDStr := qry.Get("installation_id")
-	installationID, err := strconv.Atoi(installationIDStr)
-	if err != nil || installationID == 0 {
-		http.Error(w, fmt.Sprintf("unexpected installation_id=%q", installationIDStr), http.StatusBadRequest)
-		return
-	}
 
-	// Check there's an authenticated user (this should always be the case for flows initiated by githubConnect)
 	claims := auth.GetClaims(r.Context())
 	if claims.OwnerType() != auth.OwnerTypeUser {
 		http.Error(w, "only authenticated users can connect to github", http.StatusUnauthorized)
 		return
 	}
 
-	// Associate the user with the installation
-	err = s.admin.ProcessUserGithubInstallation(r.Context(), claims.OwnerID(), int64(installationID))
+	remoteURL := qry.Get("state")
+	account, repo, ok := gitutil.SplitGithubURL(remoteURL)
+	if !ok {
+		// request without state can come in multiple ways like
+		// 	- if user changes app installation directly on the settings page
+		//  - if admin user accepts the installation request
+		// Nothing to be done in such cases. May be redirect to some rill page ??
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if setupAction == "request" {
+		// access requested
+		redirectURL, err := url.Parse(s.urls.githubConnectRequest)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create connect request url: %s", err.Error()), http.StatusInternalServerError)
+			return
+		}
+
+		qry := redirectURL.Query()
+		qry.Set("remote", remoteURL)
+		redirectURL.RawQuery = qry.Encode()
```

Comment:
As previously suggested, if we have a URL helper type, let's use it to construct URLs